homeassistant/components/caldav/calendar.py

```diff
@@ -193,27 +193,53 @@ def __init__(self, calendar, include_all_day, search):
     @Throttle(MIN_TIME_BETWEEN_UPDATES)
     def update(self):
         """Get the latest data."""
+        start_of_today = dt.start_of_local_day()
+        start_of_tomorrow = dt.start_of_local_day() + timedelta(days=1)
+
         # We have to retrieve the results for the whole day as the server
         # won't return events that have already started
-        results = self.calendar.date_search(
-            dt.start_of_local_day(), dt.start_of_local_day() + timedelta(days=1)
-        )
+        results = self.calendar.date_search(start_of_today, start_of_tomorrow)
+
+        # Create new vevents for each recurrence of an event that happens today.
+        # For recurring events, some servers return the original event with recurrence rules
+        # and they would not be properly parsed using their original start/end dates.
+        vevents = [event.instance.vevent for event in results]
+        new_vevents = []
+        for vevent in vevents:
+            rrules = vevent.getrruleset()  # Returns a list of datetimes or `None`.
+            if rrules:
+                for start_dt in rrules:
+                    today = copy.copy(start_of_today)
+                    tomorrow = copy.copy(start_of_tomorrow)
+                    if self.is_all_day(vevent):
+                        start_dt = start_dt.date()
+                        today = today.date()
+                        tomorrow = tomorrow.date()
+                    if today <= start_dt <= tomorrow:
+                        new_vevent = copy.deepcopy(vevent)
+                        if hasattr(new_vevent, "dtend"):
+                            dur = new_vevent.dtend.value - new_vevent.dtstart.value
+                            new_vevent.dtend.value = start_dt + dur
+                        new_vevent.dtstart.value = start_dt
+                        new_vevents.append(new_vevent)
+                    elif tomorrow <= start_dt:
+                        break
+            if not rrules:
```

Comment:
Put this before the if statement, then remove the if statement. You now safe an indentation on the whole code block

Reply:
Aha

Reply:
I'll do ya one better ... see the latest commit